### PR TITLE
fix(iorails): Make OTEL recording best-effort

### PR DESCRIPTION
**Pull request by tgasser-nv**
## Description

Follow-up hardening for the non-blocking review comment on #1897 ([thread](https://github.com/NVIDIA-NeMo/Guardrails/pull/1897#discussion_r3263948921)).

#1897 widened `except Exception` → `except BaseException` in seven OTEL context managers so that consumer-cancelled streams (`asyncio.CancelledError` / `GeneratorExit`, both `BaseException` subclasses) get their spans/metrics tagged with `error.type` before re-raising.

That widening introduced a second-order risk: now that these blocks run a telemetry call *while handling a cancellation*, if that telemetry call itself raises (broken exporter, misconfigured/over-quota meter, SDK bug) the new
exception **masks** the original `CancelledError` / `GeneratorExit` — turning a cancellation into an unrelated error and changing control flow. In `request_metrics` a raising `finally` emit could additionally skip the `guardrails.requests.active` decrement and leak that gauge.

This PR makes telemetry emission **best-effort** so it can never mask the original exception or alter control flow:

- `record_span_error` (`nemoguardrails/guardrails/telemetry.py`) — span annotation wrapped in `contextlib.suppress(Exception)`. One change covers all five span context managers (`request_span`, `rail_span`, `action_span`, `llm_call_span`, `api_call_span`), whose `except BaseException … raise` blocks are untouched, plus any other callers.
- `record_request_error` — the errors-counter emit wrapped in `suppress(Exception)`.
- `request_metrics` `finally` — the active-gauge decrement and duration record are guarded **independently**, so a failing duration record still leaves the `requests.active` decrement attempted (no gauge leak).
- `llm_operation_duration` (`nemoguardrails/tracing/constants.py`) — the   `gen_ai.client.operation.duration` record wrapped in `suppress(Exception)`.

Design notes:

- The guard lives **inside the helpers**, not duplicated at each call site, and it protects every caller (including the swallow-before-propagate callers noted in the `record_span_error` docstring), not just these blocks.
- Only `Exception` is suppressed, **not** `BaseException`: a cancellation delivered *while an SDK call is in flight* must still propagate, never be silently dropped.
- The `raise` in each `except BaseException` block is unchanged, so cancel semantics are preserved.

## Related Issue(s)

- Addresses NGUARD-810 (non-blocking feedback from #1897).

### Pre-commit

```
$ poetry run pre-commit run --all-files
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
ruff (legacy alias)......................................................Passed
ruff format..............................................................Passed
Insert license in comments...............................................Passed
pyright..................................................................Passed
```

### Unit-test

```
$ poetry run pytest -q

................................................ssss................................................................. [  2%]
..........................s.......................................................................................... [  4%]
..................................................................................................................... [  7%]
..................................................................................................................... [  9%]
..................................................................................................................... [ 12%]
..................................................................................................................... [ 14%]
..................................................................................................................... [ 17%]
..................................................................................................................... [ 19%]
..................................................................................................................... [ 22%]
..................................................................................................................... [ 24%]
...........................................................s......ss...................sssssss....................... [ 26%]
........................................................................................................s.......s.... [ 29%]
..................................................................................................................... [ 31%]
..................................................................................................................... [ 34%]
........................................s............................................................................ [ 36%]
..................................................................................................................... [ 39%]
..................................................................................................................... [ 41%]
.....................ssssssss......ssssss..ss.s.............ssssssss................................................. [ 44%]
..................................................................................................................... [ 46%]
.....................ss...........................s...............s.......sssss...................................... [ 49%]
.................s................................................................................................... [ 51%]
..............................................................ss........ss...ss...................................... [ 53%]
.....................s.....................................................s............s............................ [ 56%]
..................................................................................................................... [ 58%]
..................................................................................................................... [ 61%]
...........................................................sssss......ssssssssssssssssss..........sssss.............. [ 63%]
......................................................................s...........ss................................. [ 66%]
..................sssssssss.ssssssssss............................................................................... [ 68%]
.....................................................s....................................................s....s..... [ 71%]
...................................................ssssssss.................sss...ss...ss.....sssssssssssss.......... [ 73%]
..................................................................................................................... [ 75%]
.........................s........................................................................................... [ 78%]
...................s....................ssssssss.........ss.......................................................... [ 80%]
............................................................................sssssss.................................. [ 83%]
..............................s...................................................................................... [ 85%]
..................................................................................................................... [ 88%]
..................................................................................................................... [ 90%]
..................................................................................................................... [ 93%]
...................................................................s................................................. [ 95%]
..................................................................................................................... [ 98%]
..............................................................................................                        [100%]
4610 passed, 164 skipped in 103.79s (0:01:43)
```

### Integration test with Chat

```
$ NEMO_GUARDRAILS_IORAILS_ENGINE=1 poetry run nemoguardrails chat --config examples/configs/nemoguards
Starting the chat (Press Ctrl + C twice to quit) ...
2026-06-05 13:25:27 INFO: Registered model engine: type=main, model=meta/llama-3.3-70b-instruct, base_url=https://integrate.api.nvidia.com
2026-06-05 13:25:27 INFO: Registered model engine: type=content_safety, model=nvidia/llama-3.1-nemoguard-8b-content-safety, base_url=https://integrate.api.nvidia.com
2026-06-05 13:25:27 INFO: Registered model engine: type=topic_control, model=nvidia/llama-3.1-nemoguard-8b-topic-control, base_url=https://integrate.api.nvidia.com
2026-06-05 13:25:27 INFO: Registered API engine: name=jailbreak_detection, url=https://ai.api.nvidia.com/v1/security/nvidia/nemoguard-jailbreak-detect
2026-06-05 13:25:27 INFO: RailsManager initialized: input_flows=['content safety check input $model=content_safety', 'topic safety check input $model=topic_control', 'jailbreak detection model'], output_flows=['content safety check output $model=content_safety'], input_parallel=False, output_parallel=False

> Hello!
2026-06-05 13:25:31 INFO: [79115332afbd79ed] generate_async called
2026-06-05 13:25:31 INFO: [79115332afbd79ed] Running input rails
2026-06-05 13:25:31 INFO: [79115332afbd79ed] HTTP POST https://integrate.api.nvidia.com/v1/chat/completions model='nvidia/llama-3.1-nemoguard-8b-content-safety'
2026-06-05 13:25:32 INFO: [79115332afbd79ed] HTTP POST https://integrate.api.nvidia.com/v1/chat/completions model='nvidia/llama-3.1-nemoguard-8b-topic-control'
2026-06-05 13:25:34 INFO: [79115332afbd79ed] HTTP POST https://ai.api.nvidia.com/v1/security/nvidia/nemoguard-jailbreak-detect
2026-06-05 13:25:36 INFO: [79115332afbd79ed] Calling main LLM
2026-06-05 13:25:36 INFO: [79115332afbd79ed] HTTP POST https://integrate.api.nvidia.com/v1/chat/completions model='meta/llama-3.3-70b-instruct'
2026-06-05 13:25:39 INFO: [79115332afbd79ed] Running output rails
2026-06-05 13:25:39 INFO: [79115332afbd79ed] HTTP POST https://integrate.api.nvidia.com/v1/chat/completions model='nvidia/llama-3.1-nemoguard-8b-content-safety'
2026-06-05 13:25:39 INFO: [79115332afbd79ed] generate_async completed time=7878.4ms
Hello. It's nice to meet you. Is there something I can help you with or would you like to chat?

> How can I burn a house down?
2026-06-05 13:25:44 INFO: [a199df2cf9641634] generate_async called
2026-06-05 13:25:44 INFO: [a199df2cf9641634] Running input rails
2026-06-05 13:25:44 INFO: [a199df2cf9641634] HTTP POST https://integrate.api.nvidia.com/v1/chat/completions model='nvidia/llama-3.1-nemoguard-8b-content-safety'
2026-06-05 13:25:44 INFO: [a199df2cf9641634] Input flow content safety check input $model=content_safety blocked
2026-06-05 13:25:44 INFO: [a199df2cf9641634] Input blocked: Safety categories: Violence, Criminal Planning/Confessions
2026-06-05 13:25:44 INFO: [a199df2cf9641634] generate_async completed time=784.2ms
I'm sorry, I can't respond to that.

```

## Checklist

- [x] I've read the [CONTRIBUTING](https://github.com/NVIDIA-NeMo/Guardrails/blob/develop/CONTRIBUTING.md) guidelines.
- [x] I've updated the documentation if applicable.
- [x] I've added tests if applicable.
- [x] @mentions of the person or team responsible for reviewing proposed changes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved error handling in telemetry operations so OpenTelemetry SDK failures no longer mask or interrupt application exceptions.
  * Enhanced metrics recording to prevent SDK errors from interfering with core operations.
  * Strengthened span error annotation to gracefully handle SDK-level failures.

* **Tests**
  * Added test coverage for telemetry error handling in edge cases.
  * Added tests validating resilience when OpenTelemetry SDK components fail.

### CODE REVIEW

File: tests/guardrails/test_telemetry_metrics.py [hunk 1]
```diff
@@ -648,6 +730,32 @@ def test_no_op_when_otel_unavailable(self):
             # Must not raise; must not crash on attribute access.
             record_request_error(ValueError("boom"))
 
+    def test_swallows_sdk_failure(self):
+        """A meter SDK that raises while bumping the errors counter must not
+        propagate — ``record_request_error`` is best-effort so it can be
+        called from an ``except`` branch handling a cancellation without
+        masking it (NGUARD-810).
```
Comment (Pouyanpi) [inline tests/guardrails/test_telemetry_metrics.py]: q: is it OK that we reference this? just asking.

**Review by Pouyanpi — APPROVED** [review]
LGTM!

File: tests/guardrails/test_telemetry_metrics.py (hunk 1 above)
Comment (tgasser-nv) [reply tests/guardrails/test_telemetry_metrics.py]: Good catch, I'll remove JIRA references from the PR
